Initial threshold dependent interventions

pratikunterwegs opened the pull request:
This is a draft PR to get a first version of what threshold-dependent interventions could look like up and running. The main goal is to check whether this looks okay and whether it might fit the package needs in terms of extensibility (IMO, no and no).

The `daedalus()` function signature now accepts arguments for the response strategy (aka the economic configuration) and the implementation level, as well as single named values for the threshold value of a compartment that triggers the response, and a named value for the compartment that ends the response.

This works by adding a dummy switch variable to the state container which is set to 0.0 by default, and has a $dx = 0.0$, changing to 1.0 only when the response is active. This switches the scaling of the transmission parameter off (0.0) or on (1.0). The overall implementation is taken from [this SO answer by one of the {deSolve} developers](https://stackoverflow.com/a/70195166/7410366), as an initial try at implementing this using the {deSolve} events paradigm.

The basic functionality is shown in the vignette "thresholded_response.Rmd". The output does not look great for one key compartment of interest (hospitalisations) -- too spiky; but I might be wrong and this is okay.

Next steps: if this does _not_ look okay, I will move to implementing closures* as conditionals within the ODE RHS (which I am given to believe is not good practice if it introduces discontinuities, but it's mostly worked for me in [{epidemics}](https://github.com/epiverse-trace/epidemics/blob/main/inst/include/intervention.h)).

*confusing terminology but taken from EPPI parlance and referring to closing economic sectors.

richfitz commented on R/closure.R:
```diff
@@ -0,0 +1,57 @@
+  }
+
+  event_function <- function(time, state, parameters) {
+    state["switch"] <- 1.0
```
Why use a name here and not an index from the constants? Especially as it looks like above (l.11) that this is the last state, so the position is known?

pratikunterwegs replied:
I found this easier to track than the last position - I have no strong preferences either way.

pratikunterwegs replied:
This is now gone in favour of a `switch` in an `env` in `parameters`.

richfitz reviewed (commented):
Sorry, I won't have time to look at this today, though I have made a start - can we defer merging until we're both back?

It's a bit hard to review given the size, but also there are few different things all mixed in one PR (see the size of the NEWS item!).  Depending on what you're looking for/hoping for in review this will either likely need discussion or just waving through.

pratikunterwegs commented:
Thanks for taking an initial look @richfitz - I can hold off on merging then and continue doing orthogonal work on classes. The PR is indeed quite large, and I can try to explain it better once we're back.

The most useful thing would be to know whether the way I'm generating event and rootfinding functions and using them is okay (for now), or whether there's an obviously better (for any definition of better) way of doing it given what we know of the requirements.

richfitz commented:
I had hoped to spend more on it today, but it's all gone a bit pear shaped in the rush to clear the decks.

One thing that might be worth doing is splitting out some of the data processing/import bits from this PR into one uncontroversial one so that we can hone in on the interesting bits?

pratikunterwegs commented:
Sure, I can try that although it might be just two totally different PRs then. I'll close this one for now and work towards two smaller ones.

richfitz commented:
If it feels like too much work at this point, don't sweat it - talk soon!

pratikunterwegs commented:
@richfitz - I've split this PR into three, with #16 and #17 hopefully quite routine. Once those are merged, this PR should have many fewer commits, which should hopefully be easier to review. I've reordered and squashed some initial WIP commits where retaining them didn't add to the context - let me know if you'd prefer to see the full development history instead and I can force push that from my local branch.

pratikunterwegs commented:
Just reopening this for review after some reorganisation - hopefully easier to understand now; happy to discuss!

richfitz commented on R/closure.R:
```diff
@@ -0,0 +1,60 @@
+make_rt_end_event <- function() {
+  # NOTE: state reconstruction could be sped up
+  root_function <- function(time, state, parameters) {
+    state <- state[-length(state)]
```
drop now?

richfitz commented on R/closure.R:
```diff
@@ -0,0 +1,60 @@
+  event_function <- function(time, state, parameters) {
+    # prevent flipping switch when checkEventFunc runs
+    if (time != parameters[["min_time"]]) {
+      rlang::env_poke(parameters[["mutables"]], "switch", 0.0)
```
why env_poke and not
```suggestion
      parameters[["mutables"]]$switch <- 0.0
```
?

pratikunterwegs replied:
I had already added {rlang} as a dependency, and I thought it would be helpful to make it clear that one is working with an environment, rather than a list. I'm fine with moving to the base R implementation and/or removing {rlang} altogether.

richfitz commented on R/daedalus.R:
```diff
@@ -181,16 +212,91 @@ daedalus <- function(country,
+    list(
+      openness = openness,
+      mutables = mutables,
+      min_time = 1 # setting minimum time to prevent switch flipping
```
Curious why we're starting at 1 and not 0?

richfitz commented on R/daedalus.R:
```diff
@@ -181,16 +212,91 @@ daedalus <- function(country,
+  initial_state <- utils::tail(
+    data_stage_one[, colnames(data_stage_one) != "time"], 1L
+  )
+  initial_state <- c(initial_state)
```
Here are you using `c` to drop attributes of some sort? Consider `as.numeric()` or `drop()` to make that clearer?

richfitz commented on R/daedalus.R:
```diff
@@ -181,16 +212,91 @@ daedalus <- function(country,
 
-  data
+  # carry over initial state; could be named more clearly?
+  initial_state <- utils::tail(
+    data_stage_one[, colnames(data_stage_one) != "time"], 1L
+  )
```
```suggestion
  initial_state <- data_stage_one[nrow(data_stage_one), colnames(data_stage_one) != "time"]]
```
perhaps

richfitz commented on R/daedalus.R:
```diff
@@ -181,16 +212,91 @@ daedalus <- function(country,
+  rlang::env_poke(parameters[["mutables"]], "switch", 1.0)
+
+  # reset min time
+  parameters[["min_time"]] <- response_time
```
do we still need events in on the second phase? Or is this for future generality?

pratikunterwegs replied:
Yes we do, to terminate the response and lift the economic closures when Rt < 1.0.

Just thinking of future-proofing, this implementation isn't robust to a potential future need for response strategies to turn closures on and off, and we'll have to come up with a different implementation that _does_ check for at least two roots simultaneously.

richfitz commented on R/daedalus.R:
```diff
@@ -181,16 +212,91 @@ daedalus <- function(country,
+
+  data <- rbind(data_stage_one, data_stage_two)
+
+  data <- data[!duplicated(data[, "time"]), ]
```
Can you just drop the first element from stage 2 in the bind?

```
  data <- rbind(data_stage_one, data_stage_two[-1, ])
```

richfitz commented on R/prepare_output.R:
```diff
@@ -32,13 +29,14 @@ prepare_output <- function(output) {
 
   # data values
-  values <- c(output[, setdiff(colnames(output), "time")])
+  values <- c(output[, setdiff(colnames(output), c("time", "switch"))])
```
switch is gone now?

pratikunterwegs commented:
Thanks, will look at these in a bit!

pratikunterwegs commented on R/daedalus.R:
```diff
@@ -181,16 +212,80 @@ daedalus <- function(country,
+  names(initial_state) <- as.character(seq_along(initial_state))
+  initial_state <- c(initial_state, switch = 1.0)
+
+  data_stage_two <- deSolve::lsoda(
```
> we should check later if lsoda is the best bet here

I've opened an [issue for that here](https://mrc-ide.myjetbrains.com/youtrack/issue/JIDEA-91/Consider-alternatives-to-lsoda).

> we should make it easy to tweak the call here

I've just gone with the defaults for now; [another issue to pass options to the solver is here](https://mrc-ide.myjetbrains.com/youtrack/issue/JIDEA-92/Allow-passing-options-to-ODE-solver).

pratikunterwegs commented on R/ode.R:
```diff
@@ -72,63 +73,74 @@ daedalus_rhs <- function(t, state, parameters) {
+  # scaling economic sector openness
+  openness <- parameters[["openness"]]
+  switch <- state["switch"]
+  scaling <- 1 - (1 - openness) * switch # clunky
```
It would be `if (switch) openness else 1.0`; the double subtraction is just to make that logical statement work with numeric values.

richfitz reviewed (approved):
No strong feeling on rlang as a dependency, that was more query of style, and no preference at all about the start time - mostly a question of if that was capturing the intent (so running with an end time of 200 now runs for 199 days, which might be a surprise). All these things are just details though, let's merge this.
